builder/api/src/main/java/io/helidon/builder/api/ProvidedUtil.java

```diff
@@ -0,0 +1,192 @@
+                                                             Class<T> configType,
+                                                             boolean allFromServiceLoader) {
+
+        boolean discoverServices = config.get("discover-services").asBoolean().orElse(allFromServiceLoader);
+        Config providersConfig = config.get("providers");
+
+        List<ConfiguredService> configuredServices = new ArrayList<>();
+
+        // all child nodes of the current node
+        List<Config> serviceConfigList = providersConfig.asNodeList()
+                .orElseGet(List::of);
+        boolean isList = providersConfig.isList();
+
+        for (Config serviceConfig : serviceConfigList) {
+            configuredServices.add(configuredService(serviceConfig, isList));
+        }
+
+        // now we have all service configurations, we can start building up instances
+        if (config.isList()) {
+            // driven by order of declaration in config
+            return servicesFromList(serviceLoader, providerType, configType, configuredServices, discoverServices);
+        } else {
+            // driven by service loader order
+            return servicesFromObject(serviceLoader, providerType, configType, configuredServices, discoverServices);
+        }
+    }
+
+    private static <T extends NamedService> List<T>
+    servicesFromObject(HelidonServiceLoader<? extends ConfiguredProvider<T>> serviceLoader,
+                       Class<? extends ConfiguredProvider<T>> providerType,
+                       Class<T> configType,
+                       List<ConfiguredService> configuredServices,
+                       boolean allFromServiceLoader) {
+        // order is determined by service loader
+        Set<String> availableProviders = new HashSet<>();
+        Map<String, ConfiguredService> allConfigs = new HashMap<>();
+        configuredServices.forEach(it -> allConfigs.put(it.type(), it));
+        Set<String> unusedConfigs = new HashSet<>(allConfigs.keySet());
+
+        List<T> result = new ArrayList<>();
+
+        serviceLoader.forEach(provider -> {
+            ConfiguredService configuredService = allConfigs.get(provider.configKey());
+            availableProviders.add(provider.configKey());
+            unusedConfigs.remove(provider.configKey());
+            if (configuredService == null) {
+                if (allFromServiceLoader) {
+                    result.add(provider.create(Config.empty(), provider.configKey()));
+                }
+            } else {
+                if (configuredService.enabled()) {
+                    result.add(provider.create(configuredService.serviceConfig(),
+                                               configuredService.name()));
+                }
+            }
+        });
+        if (!unusedConfigs.isEmpty()) {
+            throw new ConfigException("Unknown provider configured. Expected providers with types: " + unusedConfigs
```

Comment:
why is this always a hard exception, instead of cases where we generate a info or warning but are allowed to move on? seems to aggressive in nature to me.

Reply:
Because you have explicitly defined a service you want to use, but it is not available. This is a configuration error, and silently ignoring it would cause trouble later on.

Reply:
We do the same thing in security - if you define a security provider in configuration, but do not have it configured, it is an error

Reply:
I still feel there should be a flag for this that we can toggle.

Reply:
If you find a use case in Helidon where this is correct, please add an issue and we will do it. Right now this feels like the correct "fail-fast" principle.